staging/src/k8s.io/apiserver/pkg/server/graceful_shutdown_test.go

```diff
@@ -0,0 +1,199 @@
+0tPMURSYIZ+JW7ghMspbm1XV1NTreCB/llLNqUGQ7zLAmH+KyqJK8O37/oh3VHrV
+6jp9pqrqmibtGEIpQi4D9IM8Zo9mc8GexCf0x+11mamC+ZXjT+bvLQzbcJGnG5CL
+W+S7SneWTL09leh5ATNhog6s
+-----END PRIVATE KEY-----`)
+
+// TestGracefulShutdownForActiveHTTP2Streams checks if graceful shut down of HTTP2 server works.
+// It expects that all active connections will be finished (without any errors) before the server exits.
+//
+// The test sends 25 requests to the target server in parallel. Each request is held by the target server for 60s.
+// As soon as the target server receives the last request the test calls backendServer.Config.Shutdown which gracefully shuts down the server without interrupting any active connections.
+//
+// See more at: https://github.com/golang/go/issues/39776
+//
+// Note this test will fail on upstream golang 1.15
+// OpenShift carries a patch that makes it to work
+func TestGracefulShutdownForActiveHTTP2Streams(t *testing.T) {
+	// TODO(p0lyn0mial): remove this check when upstream switches to go 1.16
+	// Note if this tests fails on 1.16 it means that the fix hasn't been provided and we should wait until the next release
+	if strings.Contains(runtime.Version(), "go1.15") {
+		t.Skip()
+	}
+	// set up the backend server
+	backendHandler := &backendHTTPHandler{}
+	backendServer := httptest.NewUnstartedServer(backendHandler)
+	backendCert, err := tls.X509KeyPair(backendCrt, backendKey)
+	if err != nil {
+		t.Fatalf("backend: invalid x509/key pair: %v", err)
+	}
+	backendServer.TLS = &tls.Config{
+		Certificates: []tls.Certificate{backendCert},
+		NextProtos:   []string{http2.NextProtoTLS},
+	}
+	backendServer.StartTLS()
+	defer backendServer.Close()
+
+	// set up the client
+	clientCACertPool := x509.NewCertPool()
+	clientCACertPool.AppendCertsFromPEM(backendCrt)
+	clientTLSConfig := &tls.Config{
+		RootCAs:    clientCACertPool,
+		NextProtos: []string{http2.NextProtoTLS},
+	}
+	client := &http.Client{}
+	client.Transport = &http2.Transport{
+		TLSClientConfig: clientTLSConfig,
+	}
+
+	// client request
+	sendRequest := func(wg *sync.WaitGroup) {
+		defer func() {
+			wg.Done()
+		}()
+
+		// act
+		resp, err := client.Get(fmt.Sprintf("https://127.0.0.1:%d", backendServer.Listener.Addr().(*net.TCPAddr).Port))
+		if err != nil {
+			t.Errorf("%v", err)
+			return
+		}
+
+		// validate
+		defer resp.Body.Close()
+		_, err = ioutil.ReadAll(resp.Body)
+		if err != nil {
+			t.Errorf("%v", err)
+		}
+		if resp.StatusCode != 200 {
+			t.Errorf("unexpected HTTP staus: %v, expected: 200", resp.StatusCode)
+		}
+		expectedProto := "HTTP/2.0"
+		if resp.Proto != expectedProto {
+			t.Errorf("unexpected response proto: %v, expected: %v", resp.Proto, expectedProto)
+		}
+	}
+
+	// this function starts the graceful shutdown
+	go func() {
+		<-startGC                   // signal from the backend after receiving all (25) requests
+		time.Sleep(5 * time.Second) // gives the last connection a chance to go into the sleep state
```

Comment:
Where does the 5 seconds come from?

Can you elaborate on why this sleep is needed? Is it to check if the server terminates connections gracefully even if the connection is in the "sleep state"?

Also I thought that HTTP2 would multiplex requests on the same TCP connection, so there's only one TCP connection throughout the test. If so, then the phrase "the last connection" is misleading.

Reply:
You are right, this sleep is unnecessary. I'm going to remove it. Thanks.